pkgs/development/compilers/terra/default.nix

```diff
@@ -1,68 +1,87 @@
-{ lib, stdenv, fetchurl, fetchFromGitHub
-, llvmPackages, ncurses, lua
-}:
+{ lib, stdenv, fetchurl, fetchFromGitHub, llvmPackages, ncurses, cmake, libxml2
+, symlinkJoin, breakpointHook, cudaPackages, enableCUDA ? false }:
 
 let
-  luajitArchive = "LuaJIT-2.0.5.tar.gz";
+  luajitRev = "9143e86498436892cb4316550be4d45b68a61224";
+  luajitArchive = "LuaJIT-${luajitRev}.tar.gz";
   luajitSrc = fetchurl {
-    url    = "http://luajit.org/download/${luajitArchive}";
-    sha256 = "0yg9q4q6v028bgh85317ykc9whgxgysp76qzaqgq55y6jy11yjw7";
+    url = "https://github.com/LuaJIT/LuaJIT/archive/${luajitRev}.tar.gz";
+    sha256 = "0kasmyk40ic4b9dwd4wixm0qk10l88ardrfimwmq36yc5dhnizmy";
   };
-in
-stdenv.mkDerivation rec {
+  llvmMerged = symlinkJoin {
+    name = "llvmClangMerged";
+    paths = with llvmPackages; [
+      llvm.out
+      llvm.dev
+      llvm.lib
+      clang-unwrapped.out
+      clang-unwrapped.dev
+      clang-unwrapped.lib
+    ];
+  };
+  cuda = cudaPackages.cudatoolkit_11;
+in stdenv.mkDerivation rec {
   pname = "terra";
-  version = "1.0.0pre1175_${builtins.substring 0 7 src.rev}";
+  version = "1.0.0-beta3";
 
   src = fetchFromGitHub {
-    owner  = "zdevito";
-    repo   = "terra";
-    rev    = "ef6a75ffee15a30f3c74f4e6943851cfbc0fec3d";
-    sha256 = "0aky17vbv3d9zng34hp17p9zb00dbzwhvzsdjzrrqvk9lmyvix0s";
+    owner = "terralang";
+    repo = "terra";
+    rev = "release-${version}";
+    sha256 = "15ik32xnwyf3g57jvvaz24f6a18lv3a86341rzjbs30kd5045qzd";
   };
 
-  nativeBuildInputs = [ lua ];
-  buildInputs = with llvmPackages; [ llvm clang-unwrapped ncurses ];
+  nativeBuildInputs = [ cmake ];
+  buildInputs = [ llvmMerged ncurses libxml2 ] ++ lib.optional enableCUDA cuda;
+
+  cmakeFlags = [
+    "-DHAS_TERRA_VERSION=0"
+    "-DTERRA_VERSION=${src.rev}"
+    "-DTERRA_LUA=luajit"
+  ] ++ lib.optional enableCUDA "-DTERRA_ENABLE_CUDA=ON";
 
   doCheck = true;
   enableParallelBuilding = true;
   hardeningDisable = [ "fortify" ];
   outputs = [ "bin" "dev" "out" "static" ];
 
-  patches = [ ./nix-cflags.patch ];
-  postPatch = ''
-    substituteInPlace Makefile \
-      --replace '-lcurses' '-lncurses'
+  patches = [
+    # Should be removed as https://github.com/terralang/terra/pull/496 get merged and released.
+    ./get-compiler-from-envvar-fix-cpu-detection.patch
+    ./nix-cflags.patch
+    ./disable-luajit-file-download.patch
+    ./nix-add-test-paths.patch
+  ];
+
+  INCLUDE_PATH = "${llvmMerged}/lib/clang/10.0.1/include";
```

Comment:
**DO NOT CHECK THIS IN**. This only serves the purpose of making all the tests pass *inside the test phase*, but attempting to use terra after it's built will always fail because these headers need to be embedded in terra using cmake. Otherwise, everyone trying to use terra would have to manually add `${llvmMerged}/lib/clang/10.0.1/include` to their INCLUDE_PATH variable, which is not what's supposed to happen. This must be fixed correctly by fixing the CLANG_RESOURCE_DIR issue.

Reply:
Your fix with the override is folded in as of my latest commit, [`c49d9ad` (#125896)](https://github.com/NixOS/nixpkgs/pull/125896/commits/c49d9adc385ccff57ef020456f3700fe6878619d).